tests/Unit/HtmlProcessor/AbstractHtmlProcessorTest.php

```diff
@@ -399,12 +417,12 @@ static function (array $dataset) {
      * Concatenates pairs of datasets (in a similar way to SQL `JOIN`) such that each new dataset consists of a 'row'
      * from a left-hand-side dataset joined with a 'row' from a right-hand-side dataset.
      *
-     * @param string[][] $leftDatasets
-     * @param string[][] $rightDatasets
+     * @param array<string, array<int, string>> $leftDatasets
+     * @param array<string, array<int, string>> $rightDatasets
```

Comment:
I think this syntax is clearer for object-like arrays, if supported by PSR-5 (which as far as I know is incomplete anyway - see https://github.com/phpDocumentor/phpDocumentor2/issues/557#issuecomment-390924081).  We had some relevant discussion recently at https://github.com/MyIntervals/emogrifier/pull/708/commits/d7d7b16eca7c7301d4022377cb4463c6d9f44c42#r324370364

Reply:
Though for arrays with string keys the proposal at https://github.com/phpDocumentor/phpDocumentor2/issues/557#issuecomment-390924081 looks better.

Reply:
When I come across docblocks like these that aren't quite specific enough for psalm, there are two choices:
1. use an array shape
2. use a more strictly-typed array (as was done here)

Some php analysis tools don't support array shapes and will consider such returns/variables/parameters as untyped (mixed, mixed[] etc.). adding an array shape generally requires adding an extra line replicating the `@param` to some degree.

In this particular case, the psalm issue was related to the possibility of non-string keys being found in the arrays, as `string[][]` is equivalent to `array<array-key, array<array-key, string>>`, where `array-key` can be either `string` or `int`, whereas the implementation uses only string keys on the outter array and int keys on the inner array (hence `array<string, array<int, string>>`).

For typed arrays vs. shaped arrays, see 23689aa vs. f3556f8

As for what to do with associative arrays, see the commits listed as referenced under #742.

Reply:
Generally, I'd like to keep the `string[]` notation over the `array<string>`/`array<int, string>` notation as long as possible because PhpStorm currently only supports the first notation. (There already is a ticket at JetBrains for this feature request.)

Reply:
> Generally, I'd like to keep the `string[]` notation over the `array<string>`/`array<int, string>` notation as long as possible because PhpStorm currently only supports the first notation. (There already is a ticket at JetBrains for this feature request.)

noted, will decorate with psalm-specific docblock tags in future.

Reply:
Can we not have both `@param` and `@psalm-param` here?

Reply:
that's the intention yes; I was going based on my experience of what types of param hints cause issues with other tools, but was not aware of phpstorm's particular issues.